Fluid Communities implementation

FerranPares opened the pull request:
As explained in the issue we created at https://github.com/igraph/igraph/issues/983, we are interested on implementing our community detection algorithm called "Fluid Communities" [1] into igraph. Fluid Communities is competitive in performance and computational cost, obtaining results close to best community detection algorithms like Multilevel, and being as fast as Label Propagation.

We already implemented the algorithm in igraph, so we open this pull request. Hope everything is fine.
Regards,
Ferran Parés

[1] https://arxiv.org/pdf/1703.09307.pdf

FerranPares commented:
AppVeyor build generates a couple of syntax errors on file **src\attributes.c**, which I have not commit at all. Really don't know if should I resolve this here, since seems not related with Fluid Communities implementation.
If someone could guide me, it would be nice.

szhorvat commented:
Is `no_of_communities` meant to be an input parameter?  If yes, why is it a pointer?  It seems to me that it should be an `igraph_integer_t` instead of an `int *`.

szhorvat commented on src/community.c:
```diff
@@ -2098,6 +2098,223 @@ int igraph_le_community_to_membership(const igraph_matrix_t *merges,
+  /* Initialize density and com_to_numvertices vectors */
+  IGRAPH_CHECK(igraph_vector_init(&density, (long int) *no_of_communities));
+  IGRAPH_FINALLY(igraph_vector_destroy, &density);
+  IGRAPH_CHECK(igraph_vector_init(&com_to_numvertices, (long int) *no_of_communities));
```
`igraph_vector_init()` is being used to initialize an `igraph_vector_int_t`. Use `igraph_vector_int_init()`.

szhorvat commented on src/community.c:
```diff
@@ -2098,6 +2098,223 @@ int igraph_le_community_to_membership(const igraph_matrix_t *merges,
+
+  igraph_vector_destroy(&node_order);
+  igraph_vector_destroy(&density);
+  igraph_vector_destroy(&com_to_numvertices);
```
Should be `igraph_vector_int_destroy()` per the above.

szhorvat commented:
I tried this function and I was wondering why it returns precisely one more community than requested when the graph is disconnected.  E.g. when asking for 4, it gives me 5.

If there are requirements on the connectedness or other properties of the graph, they should be checked by at the beginning of the function. If some of them are not met, the function should report an error and abort (using `IGRAPH_ERROR`).

ntamas commented:
Thanks Szabolcs for starting to review this PR. I was putting this aside for quite a while now due to lack of time and familiarity with the algorithm. Do you happen to know how the algorithm works from a theoretical point of view?

FerranPares commented:
I've made corresponding code changes to set no_of_communities variable as igraph_integer_t type. Tried to check everything you commented me, @szhorvat . 

Regarding the ´´´IGRAPH_ERROR´´´ checking the connectedness of the graph, I've to explore how to do that but you are right, I must do it. At function initialization i set membership of all nodes to community 0, which I treat as no community (it is necessary to do it the way algorithm works). Algorithm mathematically should not leave any node without community assigned at convergence if the graph is connected. When inputting a disconnected graph previous property is not ensured, so some nodes membership still assigned to 0, meaning no community assigned.

szhorvat commented:
@FerranPares Thanks for the update.

As a start, could you state what requirements your algorithm imposes on the graph?  Does it work on disconnected graphs?  Directed graphs?  Graphs with loop edges (`1-1`) or multiple edges (`1-2, 1-2`)?  Any other restrictions?

It should also be verified that the function handles edge cases (such as an empty graph or a single vertex).

These requirements should be stated in the documentation, and the function should verify them when appropriate (e.g. when checking that the input is correct doesn't take longer than running the algorithm).  There are multiple ways you can handle the requirements: e.g. if only undirected graphs are supported, then instead of checking that the graph is undirected, you can state in the documentation that edge directions will be ignored.

As for checking connectedness, you can use `igraph_is_connected()`.  If the algorithm requires a simple graph, you can use `igraph_is_simple()` for the check.

Regarding the `no_of_communities` argument, I still do not understand why it is a pointer. It seems to me that this is an input parameter, i.e. you use the value but never change it.  Why can't it be a simple value instead of a pointer to a value then?

FerranPares commented:
Implemented and documented requirements:
- Connected: It is a must. If unconnected throw an error. ```IGRAPH_EINVAL```
- Unweighted: I directly don't ask for weights as an input argument but I say in the docstring that function is for unweighted graphs. ```IGRAPH_EINVAL```
- Undirected: I've set up a warning message. If graph is directed, throw the warning message and treat graph as undirected.
- Loop and multiple edges: FluidC was not originally thought to work with loop edges or multiple edges. If graph is not simple then I throw an error. ```IGRAPH_EINVAL```
- Empty or single vertex graph: Those cases do not make sense. So, I throw an error on them (in fact, if no_of_nodes < no_of_communities I throw an additional error). ```IGRAPH_EINVAL```
- Additionally, I introduced an internal ```IGRAPH_ERROR``` check under ```IGRAPH_EINTERNAL``` code at the end of the function. It should never be activated, if it does means something went internally wrong.  ```IGRAPH_EINTERNAL```

```no_of_communities``` can be a simple value instead of a pointer. I'm not used to program with pointers, so I don't really know which are the main reasons for using a pointer or not. Think I start getting the point on when to use them.

szhorvat commented:
Thank you for making these fixes @FerranPares. One more thing I wanted to mention is that I noticed that you do not declare all variables at the beginning of blocks.  While most C compilers accept this, technically it is not valid ANSI C (C89). I am not an igraph maintainer, and I do not know how the maintainers feel about this. (Personally, I always put the declarations at the beginning in my pull requests, for compliance reasons, even though this hurts readability.)

@ntamas : I am not yet very familiar with how the algorithm works, I simply played with this implementation a bit.

FerranPares commented:
I just have 3 blocks of variables initialization, 2 of them at function initialization (before main loop) and 1 block inside the main loop at the very beginning. What do you exactly refer to? Do you want me to do option 1 or 2?
 - Option 1: Put first 2 blocks together at the very beginning of the function and leave the third as it is at the beginning of the main loop.
 - Option 2: Join all three blocks at the very beginning, the first two blocks and the third one corresponding to the main loop variables.

szhorvat commented:
Just to be clear: I am not an igraph maintainer, and I do not know how strict @ntamas and @gaborcsardi want to be about standards compliance.

Here's an excerpt from your code:

```
  if (igraph_is_directed(graph)) {
     IGRAPH_WARNING("Edge directions are ignored.");
   }
 
   /* Internal variables initialization */
   long int i, j, k, kv1;
   igraph_adjlist_t al;
```

This is not valid ANSI C because you are mixing declarations with code.  Declarations must be at the beginning of the block.

This is invalid:

```
{
    printf("xxx");
    int x=1;
}
```

This is valid:

```
{
    int x;
    
    printf("xxx");
    x = 1;
}
```

FerranPares commented:
@szhorvat Thanks for the clarification!

So, I wait for @ntamas or @gaborcsardi to say something about it.

ntamas commented:
> One more thing I wanted to mention is that I noticed that you do not declare all variables at the beginning of blocks. While most C compilers accept this, technically it is not valid ANSI C (C89). I am not an igraph maintainer, and I do not know how the maintainers feel about this.

I've had problems before with declarations that were not at the beginning of blocks (I think that MSVC on Windows is particularly picky about this), so I tend to move them to the top nowadays. It would be good if we could stick to this.

FerranPares commented:
Ok, I have placed all declaration blocks at the beginning of blocks to have valid ANSI C (C89) code.

I have seen that AppVeyor fails on installing an specific dependency in Python 3.4 before building up Igraph. Hope everything is alright on my PR.

FerranPares commented:
@ntamas The PR is ready waiting for you to check and merge it. No pressure on that, just a reminder.

vtraag commented on src/community.c:
```diff
@@ -2111,6 +2111,261 @@ int igraph_le_community_to_membership(const igraph_matrix_t *merges,
+ * This function implements the community detection method described in:
+ * Parés F, Gasulla DG, et. al. (2018) Fluid Communities: A Competitive,
+ * Scalable and Diverse Community Detection Algorithm. In: Complex Networks
+ * & Their Applications VI: Proceedings of Complex Networks 2017 (The Sixth 
```
The `&` should be replaced by `&amp;`, otherwise making the docs throws an error I think.

FerranPares commented on src/community.c:
```diff
@@ -2111,6 +2111,261 @@ int igraph_le_community_to_membership(const igraph_matrix_t *merges,
+
+  IGRAPH_CHECK(igraph_vector_init(&label_counters, (long int) no_of_communities));
+  IGRAPH_FINALLY(igraph_vector_destroy, &label_counters);
+  igraph_vector_null(&label_counters);
```
do you mean by "unnecessary" the `IGRAPH_CHECK`?

szhorvat replied:
No, the `IGRAPH_CHECK` is necessary.

What is unnecessary is the `igraph_vector_null`, as `igraph_vector_init` already initializes the vector to all-zeros.

FerranPares commented on src/community.c:
```diff
@@ -2111,6 +2111,261 @@ int igraph_le_community_to_membership(const igraph_matrix_t *merges,
+  IGRAPH_FINALLY(igraph_adjlist_destroy, &al);
+
+  /* Create storage space for counting distinct labels and dominant ones */
+  IGRAPH_VECTOR_INIT_FINALLY(&dominant_labels, 0);
```
I'm not sure about this. My initial idea on the initialization of `dominant_labels` was to create a vector of no size that will be able to growth at most to size `no_of_communities`.

What is the advantage of initializing it to a size that will be immediately cleared? Maybe you can help to clarify it.

szhorvat replied:
I have not looked at the details of your code, but I can answer the general question *"What is the advantage of initializing it to a size that will be immediately cleared?"*.

The size of an allocated memory block cannot usually be increased. Therefore, in order to increase the size of a vector, an entirely new memory block must be allocated and the contents of the old one must be copied over. Growing the size of a vector is slow, and it is best avoided. A typical way to do this is to pre-allocate as much storage as we expect the vector could ever need. This is why we have `igraph_vector_reserve` (and similar functions for all libraries that provide a vector datatype).

Immediately reserving as much storage for `dominant_labels` as it could ever need prevents any future reallocations.

As I said, I didn't look at the details of your code and I am not making claims about real-world performance impact.

FerranPares replied:
Thanks for clearing that up!
Then, I'm still not sure about what might be better between option A and B. If I understood correctly, both options might be kind of equivalent.

**Option A**
```suggestion
  IGRAPH_VECTOR_INIT_FINALLY(&dominant_labels, 0);
  IGRAPH_CHECK(igraph_vector_reserve(&dominant_labels, (long int) no_of_communities));
  IGRAPH_VECTOR_INIT_FINALLY(&nonzero_labels, 0);
  IGRAPH_CHECK(igraph_vector_reserve(&nonzero_labels, (long int) no_of_communities));
  ...
  /* Clear dominant_labels and nonzero_labels vectors */
  igraph_vector_clear(&dominant_labels);
  igraph_vector_clear(&nonzero_labels);
```

**Option B**
```suggestion
  IGRAPH_VECTOR_INIT_FINALLY(&dominant_labels, (long int) no_of_communities));
  IGRAPH_VECTOR_INIT_FINALLY(&nonzero_labels, (long int) no_of_communities));
  ...
  /* Clear dominant_labels and nonzero_labels vectors */
  igraph_vector_clear(&dominant_labels);
  igraph_vector_clear(&nonzero_labels);
```

szhorvat replied:
I would suggest Option B for its simplicity.

szhorvat commented:
Please do ping me before a final merge would take place. I do already have this integrated into IGraph/M and it would be easy for me to run a few tests.

FerranPares commented:
All proposed changes are done! Hope everything looks nice now for final merge. 

PD: Sorry for the rebase of all my commits. Did not expect to mess the references up :/.

vtraag commented:
@szhorvat feel free to check it further, as far as I am concerned it is good to go.

szhorvat commented:
Something seems to have gotten broken with the recent changes. It's simply not working correctly anymore. I am requesting 3 communities and it's giving me one.

szhorvat commented:
Sorry for not giving a more specific example—I can see this behaviour with the Mathematica interface. I need to look into what exactly is happening.

vtraag commented:
@szhorvat , it seemed to be working correctly for me. Perhaps something in the interface got changed? The whole PR was force-pushed because it was rebased, perhaps the Mathematica interface was still designed for the previous PR?

szhorvat commented:
I think we need an `#include "igraph_structural.h"` at the beginning of the file, otherwise my IDE complains about not knowing `igraph_is_simple()`.

szhorvat commented:
@vtraag @FerranPares Here's an example that goes wrong:

```c
#include <igraph.h>

int main() {
    igraph_t g;
    igraph_vector_t membership;
    igraph_real_t modularity;

    igraph_small(&g, 8, 0,
                 0,1, 0,2, 0,3, 1,2, 1,3, 2,3,
                 3,4,
                 4,5, 4,6, 4,7, 5,6, 5,7, 6,7,
                 -1);

    igraph_vector_init(&membership, 0);

    igraph_community_fluid_communities(&g, 2, &membership, &modularity);

    igraph_vector_print(&membership);

    igraph_vector_destroy(&membership);
    igraph_destroy(&g);

    return 0;
}
```

The typical membership vector it outputs is now `1 1 1 1 1 1 1 1` or `0 0 0 0 0 0 0 0` (i.e. a single community).

Previously, it was `0 0 0 0 1 1 1 1`.

The graph looks like this:

![image](https://user-images.githubusercontent.com/1212871/65955076-a004f080-e447-11e9-96d2-74a50db640e8.png)

szhorvat commented:
A final comment/question:

Is `community_fluid_communities` the best name? It's a little strange to repeat the word "community" twice. I don't have strong feeling about this though.

vtraag commented:
It seems something is not going well with resetting the `label_counters` in L2333-2336. Adding back in ` igraph_vector_null(&label_counters);` after L2262 solves the problem you identified @szhorvat.

szhorvat commented:
@FerranPares After fixing the above issue, could you make a few changes to the test?

 - Move to `examples/tests` (new place for tests, decided after you originally opened the PR)
 - Make the test deterministic by seeding the RNG.
 - Make the verification a bit more sensitive (so that it would catch issues like the above). E.g., check that there are _exactly_ as many clusters as expected.

I think it's okay to do this by adding more commits (so we can more easily keep track of changes).

FerranPares commented:
I have no clue right now why this is failing. I guess it has something to do with moving the test to the new folder, but I'm missing something.

vtraag commented:
The

vtraag commented:
The output log of Travis CI gives the following information:
```
/home/travis/build/igraph/igraph/examples/tests/igraph_community_fluid_communities.c: In function 'main':
/home/travis/build/igraph/igraph/examples/tests/igraph_community_fluid_communities.c:31:12: warning: unused variable 'i' [-Wunused-variable]
   long int i;
            ^
```
The test checks whether the output corresponds to the expected output, and this includes also these types of warnings.

Just delete L31 of `igraph_community_fluid_communities.c`, and you should be fine.

szhorvat commented:
Any comments on the naming? Would `igraph_community_fluid` be better than `igraph_community_fluid_communities`?

vtraag commented:
I agree that `igraph_community_fluid_communities` is a bit superfluous (pun intended), but the name of the original article is "Fluid Communities: ...", so I have no objection.

FerranPares commented:
Thanks for your help in merging this PR.

vtraag commented:
@FerranPares thanks for your perseverance, still engaging after 2 years!
